File: packages/shared_models/lib/src/lesson.dart
/// Education domain models (Phase 15/16).
///
/// Shapes mirror the backend lessons contract: `GET /lessons` returns the
/// localized list with per-user progress, `GET /lessons/{id or slug}`
/// adds the resolved `locale` and content `sections`, and
/// `POST /lessons/{id or slug}/complete` returns `lesson_id` +
/// `completed_at`.
library;

/// One content section within a localized lesson.
class LessonSection {
  const LessonSection(
      {required this.heading, required this.body, this.bullets = const []});

  final String heading;
  final String body;

  /// Optional bullet points rendered under the body text.
  final List<String> bullets;

  factory LessonSection.fromJson(Map<String, dynamic> json) {
    final heading = json['heading'];
    final body = json['body'];
    if (heading is! String || heading.isEmpty) {
      throw const FormatException('Lesson section requires a heading');
    }
    if (body is! String) {
      throw const FormatException('Lesson section requires a body');
    }
    final bullets = json['bullets'];
    return LessonSection(
      heading: heading,
      body: body,
      bullets: bullets is List
          ? bullets.map((item) => item.toString()).toList()
          : const [],
    );
  }

  Map<String, dynamic> toJson() =>
      {'heading': heading, 'body': body, 'bullets': bullets};
}

/// A lesson as returned by the API.
///
/// The list endpoint returns metadata + progress; the detail endpoint
/// additionally carries the resolved `locale` and `sections`.
class Lesson {
  const Lesson({
    required this.id,
    required this.slug,
    this.difficulty = 'beginner',
    this.category = 'media_literacy',
    this.estimatedMinutes = 5,
    this.orderIndex = 0,
    this.title,
    this.summary,
    this.completed = false,
    this.completedAt,
    this.locale,
    this.sections = const [],
  });

  final String id;
  final String slug;

  /// `beginner` | `intermediate` | `advanced`.
  final String difficulty;

  /// Curriculum category (e.g. `media_literacy`).
  final String category;

  /// Reading time estimate in minutes.
  final int estimatedMinutes;

  /// Curriculum ordering.
  final int orderIndex;

  /// Localized title (null when the lesson has no content in any
  /// fallback-chain locale).
  final String? title;

  /// Localized summary.
  final String? summary;

  /// Whether the requesting user has completed this lesson.
  final bool completed;

  /// When the user completed it, if completed.
  final DateTime? completedAt;

  /// The locale the content was resolved in (detail payload only).
  final String? locale;

  /// Content sections (detail payload only).
  final List<LessonSection> sections;

  factory Lesson.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final slug = json['slug'];
    if (id is! String || id.isEmpty) {
      throw const FormatException('Lesson requires an id');
    }
    if (slug is! String || slug.isEmpty) {
      throw const FormatException('Lesson requires a slug');
    }
    final sections = json['sections'];
    final completedAt = json['completed_at'];
    return Lesson(
      id: id,
      slug: slug,
      difficulty: json['difficulty'] as String? ?? 'beginner',
      category: json['category'] as String? ?? 'media_literacy',
      estimatedMinutes: (json['estimated_minutes'] as num?)?.toInt() ?? 5,
      orderIndex: (json['order_index'] as num?)?.toInt() ?? 0,
      title: json['title'] as String?,
      summary: json['summary'] as String?,
      completed: json['completed'] as bool? ?? false,
      completedAt:
          completedAt == null ? null : DateTime.parse(completedAt as String),
      locale: json['locale'] as String?,
      sections: sections is List
          ? sections
              .map((item) => LessonSection.fromJson(
                  Map<String, dynamic>.from(item as Map)))
              .toList()
          : const [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'slug': slug,
        'difficulty': difficulty,
        'category': category,
        'estimated_minutes': estimatedMinutes,
        'order_index': orderIndex,
        'title': title,
        'summary': summary,
        'completed': completed,
        'completed_at': completedAt?.toIso8601String(),
        'locale': locale,
        'sections': sections.map((section) => section.toJson()).toList(),
      };
}

/// The result of completing a lesson.
class LessonProgress {
  const LessonProgress({required this.lessonId, required this.completedAt});

  final String lessonId;
  final DateTime completedAt;

  factory LessonProgress.fromJson(Map<String, dynamic> json) {
    final lessonId = json['lesson_id'];
    final completedAt = json['completed_at'];
    if (lessonId is! String || lessonId.isEmpty) {
      throw const FormatException('LessonProgress requires a lesson_id');
    }
    if (completedAt is! String) {
      throw const FormatException('LessonProgress requires a completed_at');
    }
    return LessonProgress(
      lessonId: lessonId,
      completedAt: DateTime.parse(completedAt),
    );
  }

  Map<String, dynamic> toJson() => {
        'lesson_id': lessonId,
        'completed_at': completedAt.toIso8601String(),
      };
}

File: backend/app/domain/lesson.py
"""Education domain — media-literacy lessons with localized content (Phase 15).

Persisted into ``lessons`` / ``lesson_contents`` / ``lesson_progress``.
The aggregate carries the lesson metadata, its content resolved for the
best available locale in the fallback chain (ADR-0007), and the caller's
completion state so the API can render a lesson in one read.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any
from uuid import UUID


@dataclass(frozen=True)
class LessonSection:
    """One content section within a localized lesson."""

    heading: str
    body: str
    bullets: tuple[str, ...] = ()


@dataclass(frozen=True)
class LessonContent:
    """Localized lesson content resolved for one locale."""

    locale_code: str
    title: str
    summary: str
    sections: tuple[LessonSection, ...] = ()


@dataclass(frozen=True)
class Lesson:
    """A published lesson with its resolved content and progress.

    Attributes:
        id: Primary key of the ``lessons`` row.
        slug: Stable URL identifier (unique).
        difficulty: ``beginner`` | ``intermediate`` | ``advanced``.
        category: Curriculum category (e.g. ``media_literacy``).
        estimated_minutes: Reading time estimate.
        order_index: Curriculum ordering.
        published: Whether the lesson is visible through the API.
        content: Content resolved for the best available chain locale,
            or None when the lesson has no content in any chain locale.
        completed: Whether the requesting user has completed the lesson.
        completed_at: When the user completed it, if completed.
    """

    id: UUID
    slug: str
    difficulty: str = "beginner"
    category: str = "media_literacy"
    estimated_minutes: int = 5
    order_index: int = 0
    published: bool = True
    content: LessonContent | None = None
    completed: bool = False
    completed_at: datetime | None = None


@dataclass(frozen=True)
class LessonProgress:
    """A user's completion record for one lesson."""

    lesson_id: UUID
    completed_at: datetime


def sections_from_json(rows: list[dict[str, Any]]) -> tuple[LessonSection, ...]:
    """Convert persisted ``sections`` JSONB rows into domain sections.

    Unknown keys in a section object are ignored so content can evolve
    without breaking older readers.
    """
    sections: list[LessonSection] = []
    for row in rows:
        if not isinstance(row, dict):
            continue
        headings = row.get("heading")
        bodies = row.get("body")
        if not isinstance(headings, str) or not isinstance(bodies, str):
            continue
        bullets = row.get("bullets")
        if not isinstance(bullets, list):
            bullets = []
        sections.append(
            LessonSection(
                heading=headings,
                body=bodies,
                bullets=tuple(str(item) for item in bullets),
            )
        )
    return tuple(sections)
